### ISSUE TITLE
Please add support for GRLIR Blind super resolution model

### ISSUE THREAD

**Issue by zelenooki87** [chaiNNer-org/spandrel#15]
Thank you devs for hard work and implementing SRFormer and DAT in nightly build.
Could you consider implement
[https://github.com/ofsoundof/GRL-Image-Restoration](https://github.com/ofsoundof/GRL-Image-Restoration)
GRL IR blind super resolution model?
If it is possible?
Thank you very much.

**Comment by zelenooki87** [comment]
Thank you great devs, could you add support for blind image super resolution.
[bsr_grl_base.ckpt](https://github.com/ofsoundof/GRL-Image-Restoration/releases/download/v1.0.0/bsr_grl_base.ckpt)
For that particular model I am only interested. Thanks

**Comment by RunDevelopment** [comment]
> could you please add support for blind image super resolution model?
> [bsr_grl_base.ckpt](https://github.com/ofsoundof/GRL-Image-Restoration/releases/download/v1.0.0/bsr_grl_base.ckpt)

Okay, I really don't like this model. Not, GRLIR, just this particular model file. I don't like it for 2 reasons:

1. For some reason, this model includes an [`omegaconf`](https://github.com/omry/omegaconf) config. So we can't even add support for this model without adding `omegaconf` as a dependency to this project. 
    Of course, the best thing about this config is that it's not actually used by the model :)
2. This model is a checkpoint in the truest form. It contains a bunch of stuff only needed for training (e.g. the current iteration, the discriminator, and even VGG weights). This not only makes the file larger, it also makes it difficult to analyse the file. We detect models by analysing which weights are present in the file, so having a bunch of useless stuff really messes with that analysis.

So anyway, I hacked together some code to make this particular checkpoint load and then saved it as safetensors. [You can get it here.](https://drive.google.com/file/d/1JdzeTFiBVSia7PmSvr5VduwDdLnirxAG/view?usp=sharing) Not only does this file load, since it doesn't contain a bunch of useless data it's also 66% smaller (77MB).

In summary, we already support GRLIR blind SR models. Use my `.safetensors` version since `bsr_grl_base.ckpt` is stupid.

---

That all being said, I'll add the hacky code I made to get `bsr_grl_base.ckpt` to load. I won't add ``omegaconf`` as a dependency though, so you'll have to install this pip package yourself.

**Comment by zelenooki87** [comment]
Ough... What a awful results with that safetensors model.
But thanks for trying. will post examples from vapoursynthed module from holywu...
Comparations are extremely-

**Comment by zelenooki87** [comment]
https://mega.nz/folder/yBV1wQJR#4UIlT6hYVH4AmD8l1SS1fg
2 examples... randomly selected...

**Comment by zelenooki87** [comment]
Even after manual installing system python to use and omegaconf and all dependiences results are the same with safetensors model you provided to me. Will wait for your heck codes to load original ckpt file.... :D
Realy have not idea how holywu get it working from vapoursynth. By that I limmited to use it only for video files or individual frames...
Can not understand original code of author of GIRL how to run it for photos.
Thank you for effort.

**Comment by zelenooki87** [comment]
> > could you please add support for blind image super resolution model?
> > [bsr_grl_base.ckpt](https://github.com/ofsoundof/GRL-Image-Restoration/releases/download/v1.0.0/bsr_grl_base.ckpt)
> 
> Okay, I really don't like this model. Not GRLIR, just this particular model file. I don't like it for 2 reasons:
> 
> 1. For some reason, this model includes an [`omegaconf`](https://github.com/omry/omegaconf) config. So we can't even add support for this model without adding `omegaconf` as a dependency to this project.
>    Of course, the best thing about this config is that it's not actually used by the model :)
> 2. This model is a checkpoint in the truest form. It contains a bunch of stuff only needed for training (e.g. the current iteration, the discriminator, and even VGG weights). This not only makes the file larger, it also makes it difficult to analyse the file. We detect models by analysing which weights are present in the file, so having a bunch of useless stuff really messes with that analysis.
> 
> So anyway, I hacked together some code to make this particular checkpoint load and then saved it as safetensors. [You can get it here.](https://drive.google.com/file/d/1JdzeTFiBVSia7PmSvr5VduwDdLnirxAG/view?usp=sharing) Not only does this file load, since it doesn't contain a bunch of useless data it's also 66% smaller (77MB).
> 
> In summary, we already support GRLIR blind SR models. Use my `.safetensors` version since `bsr_grl_base.ckpt` is stupid.
> 
> That all being said, I'll add the hacky code I made to get `bsr_grl_base.ckpt` to load. I won't add `omegaconf` as a dependency though, so you'll have to install this pip package yourself.

Any news for properly load model? Thx

**Comment by RunDevelopment** [comment]
> https://mega.nz/folder/yBV1wQJR#4UIlT6hYVH4AmD8l1SS1fg
> Some examples...

Those were **really** helpful. Thanks to your example, I was able to see the problem.

The issue with my parameter detection. GRLIR has [a bunch of parameters](https://github.com/chaiNNer-org/spandrel/blob/main/src/spandrel/architectures/GRLIR/arch/grl.py#L231-L263) that are different from model to model. Normally, you are just supposed to know these parameters for the model you're loading (e.g. the program you used hard-codes these parameters for BSR GRL model), but we can't do that since we want to support unofficial models (=models by the community) as well. So instead of hard-coding parameters, we try to detect parameters from the model file itself. This usually works very well, but it's not perfect. Unfortunately, some GRLIR parameters *cannot be detected* from the model file alone, so I just left them as the default value (e.g. 8 for window size). Those default values worked well enough to *load* models, but the *output* of the models was all wrong. This is what you saw.

So I added a heuristic to guess the undetectable parameters. This works well enough to get the correct output for BSR and SR models. Thanks to your example images, I was able to verify that the output is correct now.

Unfortunately, those heuristics do not work for 1x models. So all GRLIR denoise, deblur, jpeg, etc. models are unable right now. They will load fine, but the output will be all wrong.

> Can not understand original code of author of GIRL how to run it for photos.

Same. That repo is a mess...

**Comment by zelenooki87** [comment]
Yeah, repo is a mess but results I have got with Blind model through Selurs Hybrid are awesome... Even better than commercial products.. but not easy to deal with tiles for bigger inputs... if you need more examples, please let me know.
I see you stated that output are now correct? How to test it in chaainner?
Or you will update codes later?
thanks

**Comment by zelenooki87** [comment]
> > https://mega.nz/folder/yBV1wQJR#4UIlT6hYVH4AmD8l1SS1fg
> > Some examples...
> 
> Those were **really** helpful. Thanks to your example, I was able to see the problem.
> 
> The issue with my parameter detection. GRLIR has [a bunch of parameters](https://github.com/chaiNNer-org/spandrel/blob/main/src/spandrel/architectures/GRLIR/arch/grl.py#L231-L263) that are different from model to model. Normally, you are just supposed to know these parameters for the model you're loading (e.g. the program you used hard-codes these parameters for BSR GRL model), but we can't do that since we want to support unofficial models (=models by the community) as well. So instead of hard-coding parameters, we try to detect parameters from the model file itself. This usually works very well, but it's not perfect. Unfortunately, some GRLIR parameters _cannot be detected_ from the model file alone, so I just left them as the default value (e.g. 8 for window size). Those default values worked well enough to _load_ models, but the _output_ of the models was all wrong. This is what you saw.
> 
> So I added a heuristic to guess the undetectable parameters. This works well enough to get the correct output for BSR and SR models. Thanks to your example images, I was able to verify that the output is correct now.
> 
> Unfortunately, those heuristics do not work for 1x models. So all GRLIR denoise, deblur, jpeg, etc. models are unable right now. They will load fine, but the output will be all wrong.
> 
> > Can not understand original code of author of GIRL how to run it for photos.
> 
> Same. That repo is a mess...

Maybe this would be helpful?
https://github.com/HolyWu/vs-grlir/blob/master/vsgrlir/__init__.py
